src/linear_algebra/cayley_hamilton.lean

```diff
@@ -0,0 +1,98 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import tactic.apply_fun
+import ring_theory.matrix_algebra
+import ring_theory.polynomial_algebra
+import linear_algebra.nonsingular_inverse
+import tactic.squeeze
+
+/-!
+# The Cayley-Hamilton theorem, over a commutative ring.
+
+We use a nice proof from http://drorbn.net/AcademicPensieve/2015-12/CayleyHamilton.pdf
+-/
+
+noncomputable theory
+
+universes u v w
+
+open polynomial matrix
+open_locale big_operators
+
+variables {R : Type u} [comm_ring R]
+variables {n : Type w} [fintype n] [decidable_eq n]
+
+open finset
+
+/--
+The "characteristic matrix" of `M : matrix n n R` is the matrix of polynomials $t I - M$.
+The determinant of this matrix is the characteristic polynomial.
+-/
+def char_matrix (M : matrix n n R) : matrix n n (polynomial R) :=
+matrix.scalar n (X : polynomial R) - (C : R →+* polynomial R).map_matrix M
+
+@[simp] lemma char_matrix_apply_eq (M : matrix n n R) (i : n) :
+  char_matrix M i i = (X : polynomial R) - C (M i i) :=
+by simp only [char_matrix, sub_left_inj, pi.sub_apply, scalar_apply_eq,
+  ring_hom.map_matrix_apply, map_apply]
+
+@[simp] lemma char_matrix_apply_ne (M : matrix n n R) (i j : n) (h : i ≠ j) :
+  char_matrix M i j = - C (M i j) :=
+by simp only [char_matrix, pi.sub_apply, scalar_apply_ne _ _ _ h, zero_sub,
+  ring_hom.map_matrix_apply, map_apply]
+
+lemma mat_poly_equiv_char_matrix (M : matrix n n R) :
+  mat_poly_equiv (char_matrix M) = X - C M :=
+begin
+  ext k i j,
+  simp only [mat_poly_equiv_coeff_apply, coeff_sub, pi.sub_apply],
+  by_cases h : i = j,
+  { subst h, rw [char_matrix_apply_eq, coeff_sub],
+    simp only [coeff_X, coeff_C],
+    split_ifs; simp, },
+  { rw [char_matrix_apply_ne _ _ _ h, coeff_X, coeff_neg, coeff_C, coeff_C],
+    split_ifs; simp [h], }
+end
+
+/--
+The characteristic polynomial of a matrix `M` is given by $\det (t I - M)$.
+-/
+def char_poly (M : matrix n n R) : polynomial R :=
+(char_matrix M).det
+
+/--
+The Cayley-Hamilton theorem, that the characteristic polynomial of a matrix,
+applied to the matrix itself, is zero.
+
+This holds over any commutative ring.
+-/
+-- This proof follows http://drorbn.net/AcademicPensieve/2015-12/CayleyHamilton.pdf
+theorem cayley_hamilton (M : matrix n n R) :
```

Comment:
Should this use `eval2` or `aeval`?
Should the name be `char_poly_eval2_self`? (I know, it looks ugly...)

Reply:
It used to use eval2, but then the very first step was to rewrite it back to `(_.map _).eval _`... I'll have another look.

Reply:
Yeah, I'm unconvinced that it is clearer or easier with `eval2`. The proof really does use `(_.map _).eval _`, so in order to change the theorem statement I'll need to add a penultimate line `simp only [eval₂_eq_eval_map],`

Reply:
I've renamed it to `char_poly_map_eval_self`.